src/platform/backends/shared/base_availability_zone_manager.cpp

```diff
@@ -0,0 +1,170 @@
+constexpr auto automatic_zone_key = "automatic_zone";
+} // namespace
+
+namespace mpl = multipass::logging;
+
+namespace multipass
+{
+auto create_default_zones(const fs::path& zones_directory)
+{
+    std::array<AvailabilityZone::UPtr, default_zone_names.size()> zones{};
+    std::transform(default_zone_names.begin(), default_zone_names.end(), zones.begin(), [&](const auto& zone_name) {
+        return std::make_unique<BaseAvailabilityZone>(zone_name, zones_directory);
+    });
+    return zones;
+};
+
+BaseAvailabilityZoneManager::data BaseAvailabilityZoneManager::read_from_file(
+    const std::filesystem::path& file_path,
+    const std::filesystem::path& zones_directory)
+{
+    mpl::debug(category, "reading AZ manager from file '{}'", file_path);
+
+    const auto read_json = [&] {
+        try
+        {
+            return MP_JSONUTILS.read_object_from_file(file_path);
+        }
+        catch (std::exception& e)
+        {
+            mpl::warn(category, "failed to read AZ manager file': {}", e.what());
+            return QJsonObject{};
+        }
+    };
+
+    const auto json = read_json();
+
+    const auto deserialize_automatic_zone = [&] {
+        const auto json_automatic_zone = json[automatic_zone_key].toString().toStdString();
+        if (std::find(default_zone_names.begin(), default_zone_names.end(), json_automatic_zone) !=
+            default_zone_names.end())
+            return json_automatic_zone;
+
+        mpl::debug(category, "automatic zone '{}' not known, using default", json_automatic_zone);
+        return std::string{*default_zone_names.begin()};
+    };
+
+    return {
+        // TODO remove these comments in C++20
+        /*.file_path = */ file_path,
+        /*.zone_collection = */ {create_default_zones(zones_directory), deserialize_automatic_zone()},
+    };
+}
+
+BaseAvailabilityZoneManager::BaseAvailabilityZoneManager(const fs::path& data_dir)
+    : m{read_from_file(data_dir / az_file, data_dir / zones_directory_name)}
+{
+    serialize();
+}
+
+AvailabilityZone& BaseAvailabilityZoneManager::get_zone(const std::string& name)
+{
+    for (const auto& zone : m.zone_collection.zones)
+    {
+        if (zone->get_name() == name)
+            return *zone;
+    }
+    throw AvailabilityZoneNotFound{name};
+}
+
+std::string BaseAvailabilityZoneManager::get_automatic_zone_name()
+{
+    const auto zone_name = m.zone_collection.next_available();
+    serialize();
+    return zone_name;
+}
+
+std::vector<std::reference_wrapper<const AvailabilityZone>> BaseAvailabilityZoneManager::get_zones()
+{
+    std::vector<std::reference_wrapper<const AvailabilityZone>> zone_list;
+    zone_list.reserve(m.zone_collection.zones.size());
+    for (auto& zone : m.zone_collection.zones)
+        zone_list.emplace_back(*zone);
+    return zone_list;
+}
+
+std::string BaseAvailabilityZoneManager::get_default_zone_name() const
+{
+    return m.zone_collection.zones.begin()->get()->get_name();
```

Comment:
It feels like "m.zone_collection.zones" could use a shortcut in the BaseAvailabliltyZoneManager itself:

zones() {return m.zone_collection.zones;}

Also, instead of ptr->get->get_name(), we can deref the iterator to shorten the "train":

```c++
 (*m.zone_collection.zones.begin())->get_name();
 ```